// @flatfile/adapter/build/main/streamed-results.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamedResults = void 0;
var StreamedResults = /** @class */ (function () {
    function StreamedResults(data, meta) {
        this.$meta = meta;
        this.$data = data;
        this.rawOutput = this.$data;
        this.validData = this.$data
            .filter(function (v) { return v.valid; })
            .filter(function (v) { return !v.deleted; })
            .map(function (v) { return v.data; });
        this.data = this.validData;
        this.deletedData = this.$data.filter(function (v) { return v.deleted; }).map(function (v) { return v.data; });
        this.allData = this.$data.map(function (v) { return v.data; });
        this.remainingChunks = Math.ceil((this.totalChunks - this.currentChunk) / this.$meta.inChunks);
        this.totalChunks = Math.ceil(this.$meta.count_rows_accepted / this.$meta.inChunks);
        this.chunkSize = this.$meta.inChunks;
        this.currentChunk = (this.$meta.pointer + this.chunkSize) / this.chunkSize;
        this.hasMore = this.$meta.hasMore;
    }
    return StreamedResults;
}());
exports.StreamedResults = StreamedResults;
