// @flatfile/adapter/build/module/results.js
import { Stats } from './stats';
import { EndUser } from './user';
import { UploadFile } from './upload-file';
import { StreamedResults } from './streamed-results';
var Results = /** @class */ (function () {
    function Results(data, meta, importer) {
        var _this = this;
        this.$meta = meta;
        this.$data = data;
        this.$importer = importer;
        this.rawOutput = this.blobOnly(this.$data, 'rawOutput');
        this.validData = this.blobOnly(this.$data
            .filter(function (v) { return v.valid; })
            .filter(function (v) { return !v.deleted; })
            .map(function (v) { return v.data; }), 'validData');
        this.data = this.blobOnly(this.validData, 'data');
        this.deletedData = this.blobOnly(this.$data.filter(function (v) { return v.deleted; }).map(function (v) { return v.data; }), 'deletedData');
        this.allData = this.blobOnly(this.$data.map(function (v) { return v.data; }), 'allData');
        this.batchId = this.$meta.batchID;
        this.stats = new Stats(this.$meta);
        this.customer = this.$meta.endUser ? new EndUser(this.$meta.endUser) : null;
        this.originalFile = this.$meta.originalFile ? new UploadFile(this.$meta.originalFile) : null;
        this.csvFile = this.getCSVFile();
        this.fileName = this.$meta.filename || null;
        this.managed = this.$meta.managed || false;
        this.manual = this.$meta.manual;
        this.config = this.$meta.config;
        this.parsingConfig = this.$meta.parsing_config;
        this.skippedRows = this.$meta.skipped_rows || null;
        this.headersRaw = this.$meta.headers_raw || null;
        this.headersMatched = this.$meta.headers_matched || null;
        this.customColumns = this.$meta.custom_columns;
        this.categoryFieldMap = this.$meta.category_field_map || null;
        this.failureReason = this.$meta.failure_reason || null;
        this.submittedAt = this.$meta.submitted_at || null;
        this.failedAt = this.$meta.failed_at || null;
        this.createdAt = this.$meta.created_at;
        this.stylesheet = this.$meta.stylesheet;
        this.nextChunk = function () { return _this.getNextChunk(); };
    }
    Results.prototype.getCSVFile = function () {
        if (this.$meta.originalFile) {
            if (this.$meta.originalFile.filetype === 'csv') {
                return new UploadFile(this.$meta.originalFile);
            }
            else {
                if (this.$meta.csvFile) {
                    return new UploadFile(this.$meta.csvFile);
                }
            }
        }
        return null;
    };
    Results.prototype.getNextChunk = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.$meta.inChunks) {
                return reject("\"nextChunk()\" is only accessible when using \"inChunks\". Please see docs for \"requestDataFromUser\".");
            }
            _this.$importer.$ready.then(function (child) {
                console.log('child.nextChunk()');
                child.nextChunk().then(function (data) {
                    console.log('nextChunk()', data);
                    resolve(data.results.length ? new StreamedResults(data.results, data.meta) : null);
                }, function (err) {
                    console.log('nextChunk(err)', err);
                });
            });
        });
    };
    Results.prototype.blobOnly = function (v, method, alt) {
        if (alt === void 0) { alt = 'nextChunk()'; }
        if (this.$meta.inChunks) {
            throw new Error("\"" + method + "\" is not accessible when using \"inChunks\". Please see docs for \"" + alt + "\" instead.");
        }
        return v;
    };
    return Results;
}());
export { Results };
